File: packages/service/src/domain/auth/auth-service.ts
import { RequestSession } from '@service/api/request-session';
import { User } from '@service/domain/users/user';
import { UsersService } from '@service/domain/users/users-service';
import { Logger } from '@service/infrastructure/logger';
import { Service } from 'typedi';

@Service()
class AuthService {
  private static logger = new Logger('auth-service');

  constructor(private usersService: UsersService) {}

  async authorize(username: string, password: string, session: RequestSession): Promise<void> {
    const userExists = await this.usersService.doesUserExist(username);

    if (!userExists) {
      const userCount = (await this.usersService.getAllUserData()).length;

      if (userCount === 0) {
        await this.usersService.addNewUser(username);
      } else {
        AuthService.logger.info(`User "${username}" tried to log in, but does not exist`);
        throw new Error('User does not exist');
      }
    }

    const user = await this.usersService.getByName(username);
    if (!user) throw new Error('Something went wrong');

    const userDidSetPassword: boolean = (!!user.secret);

    if (userDidSetPassword) {
      const isPasswordCorrect = await UsersService.checkPassword(password, user);
      if (isPasswordCorrect) {
        await this.loginCorrectPassword(user, session);
      } else {
        this.loginWrongPassword(user);
      }
    } else {
      await this.loginPasswordNotSet(user, password, session);
    }
  }

  async isRequestAuthorized(session: RequestSession, token: (string | null)): Promise<boolean> {
    const isSessionAuthorized: boolean = await this.isSessionAuthorized(session);
    const isApiTokenAuthorized: boolean = token
      ? await this.isApiTokenAuthorized(token)
      : false;

    return (isSessionAuthorized || isApiTokenAuthorized);
  }

  async getRequestsUser(session: RequestSession, token: (string | null)): Promise<User| null> {
    const userFromSession = await this.getUserFromSession(session);
    const userFromToken = token ? await this.usersService.getByApiToken(token) : null;
    return userFromSession || userFromToken;
  }

  async isWebSocketAuthorized(session: RequestSession): Promise<boolean> {
    return this.isSessionAuthorized(session);
  }

  private async getUserFromSession(session: RequestSession): Promise<User | null> {
    return session.loggedUser
      ? this.usersService.getByName(session.loggedUser.name)
      : null;
  }

  private async isSessionAuthorized(requestSession: RequestSession): Promise<boolean> {
    const user = await this.getUserFromSession(requestSession);
    return (user !== null);
  }

  private async isApiTokenAuthorized(token: string): Promise<boolean> {
    const user = await this.usersService.getByApiToken(token);
    return (user !== null);
  }

  private async loginCorrectPassword(user: User, session: RequestSession): Promise<void> {
    // Authorize session
    session.loggedUser = { // eslint-disable-line no-param-reassign
      name: user.data.name,
      apiToken: user.secret!.apiToken,
    };

    await this.usersService.updateLastLoginDate(user);

    AuthService.logger.info(`User "${user.data.name}" logged in`);
  }

  private loginWrongPassword(user: User): void {
    AuthService.logger.info(`User "${user.data.name}" tried to log in, but provided wrong password`);
    throw new Error('Incorrect password');
  }

  private async loginPasswordNotSet(user: User, nextPassword: string, session: RequestSession): Promise<void> {
    const userWithPassword = await this.usersService.setPassword(nextPassword, user);
    AuthService.logger.info(`User "${user.data.name}" set new password`);

    await this.loginCorrectPassword(userWithPassword, session);
  }
}

export { AuthService };

File: service/src/main/kotlin/xyz/lebkuchenfm/domain/auth/AuthService.kt
package xyz.lebkuchenfm.domain.auth

import com.github.michaelbull.result.Err
import com.github.michaelbull.result.Ok
import com.github.michaelbull.result.Result
import com.github.michaelbull.result.map
import com.github.michaelbull.result.mapError
import io.github.oshai.kotlinlogging.KotlinLogging
import xyz.lebkuchenfm.domain.sessions.UserSession
import xyz.lebkuchenfm.domain.users.CreateFirstUserError
import xyz.lebkuchenfm.domain.users.SetPasswordError
import xyz.lebkuchenfm.domain.users.UsersService

private val logger = KotlinLogging.logger {}

class AuthService(private val usersService: UsersService) {
    suspend fun authenticateWithCredentials(username: String, password: String): Result<UserSession, AuthError> {
        val user = usersService.getByName(username)

        return when {
            user == null -> {
                usersService.createFirstUser(username, password)
                    .map {
                        logger.info { "User '$username' is the first user to log in." }
                        UserSession(it.data.name, it.effectiveScopes)
                    }
                    .mapError {
                        when (it) {
                            CreateFirstUserError.UsersAlreadyExist -> {
                                logger.info { "User '$username' tried to log in, but does not exist." }
                                AuthError.UserDoesNotExistError
                            }
                            is CreateFirstUserError.PasswordValidationError -> {
                                AuthError.CannotSetPasswordError(it.error)
                            }
                            CreateFirstUserError.UnknownError -> {
                                logger.error { "Something went wrong while creating first user '$username'." }
                                AuthError.UnknownError
                            }
                        }
                    }
            }

            !user.hasPasswordSet -> {
                logger.info { "User '${user.data.name}' is logging in for the first time." }
                usersService.setPassword(user, password)
                    .map { UserSession(it.data.name, it.effectiveScopes) }
                    .mapError { AuthError.CannotSetPasswordError(it) }
            }

            else -> {
                if (usersService.checkPassword(user, password)) {
                    logger.info { "User '${user.data.name}' logged in." }
                    usersService.updateLastLoginDate(user)
                    Ok(UserSession(user.data.name, user.effectiveScopes))
                } else {
                    logger.info { "User '${user.data.name}' tried to log in, but provided wrong password." }
                    Err(AuthError.BadCredentialsError)
                }
            }
        }
    }

    suspend fun authenticateWithApiToken(token: String): UserSession? {
        return usersService.getByApiToken(token)?.let { UserSession(it.data.name, it.effectiveScopes) }
    }
}

sealed class AuthError {
    data class CannotSetPasswordError(val setPasswordError: SetPasswordError) : AuthError()
    data object UserDoesNotExistError : AuthError()
    data object BadCredentialsError : AuthError()
    data object UnknownError : AuthError()
}
